common/src/main/java/io/netty/util/internal/CleanerJava6.java

```diff
@@ -32,55 +34,112 @@
 
     static {
-        long fieldOffset = -1;
-        Method clean = null;
+        long fieldOffset;
+        Method clean;
+        Field cleanerField;
         Throwable error = null;
-        if (PlatformDependent0.hasUnsafe()) {
-            ByteBuffer direct = ByteBuffer.allocateDirect(1);
-            try {
-                Field cleanerField = direct.getClass().getDeclaredField("cleaner");
+        final ByteBuffer direct = ByteBuffer.allocateDirect(1);
+        try {
+            Object cleaner;
+            Object mayBeCleanerField = AccessController.doPrivileged(new PrivilegedAction<Object>() {
+                @Override
+                public Object run() {
+                    try {
+                        return direct.getClass().getDeclaredField("cleaner");
+                    } catch (Throwable cause) {
+                        return cause;
+                    }
+                }
+            });
+            if (mayBeCleanerField instanceof Throwable) {
+                throw (Throwable) mayBeCleanerField;
+            }
+
+            cleanerField = (Field) mayBeCleanerField;
+
+            // If we have sun.misc.Unsafe we will use it as its faster then using reflection and we do not need
+            // to open up access levels, otherwise let us try reflection as last resort.
+            if (PlatformDependent.hasUnsafe()) {
                 fieldOffset = PlatformDependent0.objectFieldOffset(cleanerField);
-                Object cleaner = PlatformDependent0.getObject(direct, fieldOffset);
-                clean = cleaner.getClass().getDeclaredMethod("clean");
-                clean.invoke(cleaner);
-            } catch (Throwable t) {
-                // We don't have ByteBuffer.cleaner().
+                cleaner = PlatformDependent0.getObject(direct, fieldOffset);
+            } else {
                 fieldOffset = -1;
-                clean = null;
-                error = t;
+                cleanerField.setAccessible(true);
+                cleaner = cleanerField.get(direct);
             }
-        } else {
-            error = new UnsupportedOperationException("sun.misc.Unsafe unavailable");
+            clean = cleaner.getClass().getDeclaredMethod("clean");
+            clean.invoke(cleaner);
+        } catch (Throwable t) {
+            // We don't have ByteBuffer.cleaner().
+            fieldOffset = -1;
+            clean = null;
+            error = t;
+            cleanerField = null;
         }
+
         if (error == null) {
             logger.debug("java.nio.ByteBuffer.cleaner(): available");
         } else {
             logger.debug("java.nio.ByteBuffer.cleaner(): unavailable", error);
         }
+        CLEANER_FIELD = cleanerField;
         CLEANER_FIELD_OFFSET = fieldOffset;
         CLEAN_METHOD = clean;
     }
 
     static boolean isSupported() {
-        return CLEANER_FIELD_OFFSET != -1;
+        return CLEANER_FIELD_OFFSET != -1 || CLEANER_FIELD != null;
```

Comment:
Note: CLEANER_FIELD != null even if using unsafe.

Reply:
yeah sure but does it matter ?

Reply:
I'd agree it doesn't really matter. But I just wanted to point it out.